BUG: lzma is a required part of python, make it optional

guilherme-salome opened the pull request:
Importing `lzma` when Python has been compiled without its support (usually due to a lack of `xz` in the system) will raise a warning. Substituted `import lzma` for helper function.

- [x] closes #27575
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

guilherme-salome committed:
Importing lzma when Python has been compiled without its support will
raise a warning. Substituted import lzma for helper function.

guilherme-salome committed:
Substituted import lzma to call to import_lzma

TomAugspurger commented on pandas/compat/__init__.py:
```diff
@@ -65,3 +65,16 @@ def is_platform_mac():
+
+
+def import_lzma():
+    import warnings
```
Can put this import at the top.

Salompas replied:
Will move `import warnings` to the top. Thanks!

Salompas commented:
I cannot modify the `pandas/_libs/parsers.pyx` without running into issues with `cythonize` (via `pip install -e .`).

TomAugspurger commented on pandas/compat/__init__.py:
```diff
@@ -65,3 +65,16 @@ def is_platform_mac():
+    except ImportError:
+        msg = (
+            "Could not import the lzma module. Your installed Python is incomplete. "
+            "Attempting to use `lzma` compression will result in a RuntimeError."
```
How strong do we want to be in wording this warning message? I'm in favor of being quite explicit that the user's python is broken. So perhaps add something like "We recommend that you reinstall Python." though we can't really give guidance on what exactly went wrong with their previous install.

Salompas replied:
Agree.
I am not sure if it would be too much to add, but if we know that this often happens due to a lack of `xz` being installed (on Mac, for example), we might want to mention that to the user.

guilherme-salome committed:
Calls to lzma module will raise a RuntimeError.

TomAugspurger reviewed (commented):
Thanks @Salompas.

We'll also want

1. A release note in 0.25.1.rst. Can go under the "IO" bug fixes section.
2. Update the uses of lzma to raise a RuntimeError when a broken Python install tries to use lzma
3. A test

For the test, something like

```diff
diff --git a/pandas/tests/io/test_compression.py b/pandas/tests/io/test_compression.py
index ce459ab24..0d359dd72 100644
--- a/pandas/tests/io/test_compression.py
+++ b/pandas/tests/io/test_compression.py
@@ -1,5 +1,7 @@
 import contextlib
 import os
+import subprocess
+import textwrap
 import warnings

 import pytest
@@ -125,3 +127,15 @@ def test_compression_warning(compression_only):
         with tm.assert_produces_warning(RuntimeWarning, check_stacklevel=False):
             with f:
                 df.to_csv(f, compression=compression_only)
+
+
+def test_with_missing_lzma():
+    # https://github.com/pandas-dev/pandas/issues/27575
+    code = textwrap.dedent(
+        """\
+    import sys
+    sys.modules['lzma'] = None
+    import pandas
+    """
+    )
+    subprocess.check_output(["python", "-c", code])
```

TomAugspurger commented:
> I cannot modify the pandas/_libs/parsers.pyx without running into issues with cythonize (via pip install -e .).

What issues do you see? You'll need a C compiler, since you're updating a Cython file. Typically I use `python setup.py build_ext --inplace`, but `pip install -e .` may do the same.

guilherme-salome committed:
Formatted with black.

Salompas commented:
> > I cannot modify the pandas/_libs/parsers.pyx without running into issues with cythonize (via pip install -e .).
> 
> What issues do you see? You'll need a C compiler, since you're updating a Cython file. Typically I use `python setup.py build_ext --inplace`, but `pip install -e .` may do the same.

I was adding an `if lzma is None` to one of the lines to see if `RuntimeError` is necessary.
However, when I run `pip install -e .` after modifying the file, I get:
```
    ERROR: Complete output from command python setup.py egg_info:
    ERROR: 
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
    
    
    cdef extern from "parser/tokenizer.h":
    
        ctypedef enum ParserState:
    	START_RECORD
    ^
    ------------------------------------------------------------
    
    pandas/_libs/parsers.pyx:90:0: Mixed use of tabs and spaces
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/guilhermesalome/Desktop/pandas/setup.py", line 813, in <module>
        ext_modules=maybe_cythonize(extensions, compiler_directives=directives),
      File "/Users/guilhermesalome/Desktop/pandas/setup.py", line 542, in maybe_cythonize
        return cythonize(extensions, *args, **kwargs)
      File "/Users/guilhermesalome/.pyenv/versions/miniconda3-latest/envs/pandas-dev/lib/python3.7/site-packages/Cython/Build/Dependencies.py", line 1096, in cythonize
        cythonize_one(*args)
      File "/Users/guilhermesalome/.pyenv/versions/miniconda3-latest/envs/pandas-dev/lib/python3.7/site-packages/Cython/Build/Dependencies.py", line 1219, in cythonize_one
        raise CompileError(None, pyx_file)
    Cython.Compiler.Errors.CompileError: pandas/_libs/parsers.pyx
    Compiling pandas/_libs/parsers.pyx because it changed.
    [1/1] Cythonizing pandas/_libs/parsers.pyx
    ----------------------------------------
ERROR: Command "python setup.py egg_info" failed with error code 1 in /Users/guilhermesalome/Desktop/pandas/
```
I am modifying the file in Emacs without any mode for Cython.

Salompas commented:
I managed to make it work in `nano` and putting whitespaces by hand. It might be an issue with my Emacs using tabs.

guilherme-salome committed:
Raise RuntimeError when calling a method of lzma when lzma is not available.

guilherme-salome committed:
Release not explaining solution to #27575.

guilherme-salome committed:
Moved import warnings to top.

Salompas commented:
> ```diff
> # https://github.com/pandas-dev/pandas/issues/27575
> ```

This is much smarter than what I was doing to test the code (change the name of the `lzma.py` module).

guilherme-salome committed:
Added test for import lzma. Test passes when lzma is not available and
you try importing it.

Salompas commented:
I do not understand what the command `git diff upstream/master -u -- "*.py" | flake8 --diff` does, but if I do `git diff master -u -- "*.py" | flake8 --diff" the output is nothing.

guilherme-salome committed:
Improved explanation of solution to #27575.

Salompas commented:
Please let me know if this is looking good, and how I can improve the solution.

guilherme-salome committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into lzma-compression-optional

jbrockmendel commented:
>  but if I do git diff master -u -- "*.py" | flake8 --diff" the output is nothing

That means it is passing that check.

The isort check is failing.  Should be fixed if you run `isort -rc pandas`

TomAugspurger commented:
@Salompas it's an isort issue: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=15881&view=logs&jobId=435ca956-9126-505f-566f-fff31072e2ba&taskId=e3717827-2a2f-5148-8ad4-7900d2171f6e&lineStart=37&lineEnd=38&colStart=1&colEnd=1

https://dev.pandas.io/development/contributing.html#import-formatting

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -103,8 +103,10 @@ MultiIndex
 I/O
 ^^^
-
+- Bug: importing `lzma` can fail when Python is not properly installed. For example, on MacOS, installing Python with `pyenv` when `xz` is not available does not raise any issues during installation. However, trying to `import pandas` afterwards will fail, since the module `lzma` will not be available. Usually, this can be solved by installing the appropriate Python dependencies, and then resinta ...
```
A bit long for a list item entry I think. This can probably be split into it's own section so we can refer to it when people ask about the warning message. So I would put this as a new section, at the same level as "Bug Fixes". Probably just be for bug fixes.

On the message, let's write it to convey

- What the change was ("Pandas will now warn, rather than raising an ImportError, if the lzma module is not present"
- Describe the issue ("Some users may unknowingly have an incomplete Python installation")
- Describe a possible fix ("Ensure you have the necessary libraries and reinstall Python").

I don't necessarily think we should be telling people how to properly use pyenv and related tools, though.

Salompas replied:
Got it.
I will create a new section at the same level as `Bug Fixes` and make the changes you suggested.

Salompas replied:
> A bit long for a list item entry I think. This can probably be split into it's own section so we can refer to it when people ask about the warning message. So I would put this as a new section, at the same level as "Bug Fixes". Probably just be for bug fixes.
> 
> On the message, let's write it to convey
> 
> * What the change was ("Pandas will now warn, rather than raising an ImportError, if the lzma module is not present"
> * Describe the issue ("Some users may unknowingly have an incomplete Python installation")
> * Describe a possible fix ("Ensure you have the necessary libraries and reinstall Python").
> 
> I don't necessarily think we should be telling people how to properly use pyenv and related tools, though.

I made the change, could you take a look if the sectioning is ok?

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -103,8 +103,10 @@ MultiIndex
+- Bug: importing `lzma` can fail when Python is not properly installed. For example, on MacOS, installing Python with `pyenv` when `xz` is not available does not raise any issues during installation. However, trying to `import pandas` afterwards will fail, since the module `lzma` will not be available. Usually, this can be solved by installing the appropriate Python dependencies, and then resinta ...
+-
 - Avoid calling ``S3File.s3`` when reading parquet, as this was removed in s3fs version 0.3.0 (:issue:`27756`)
+>>>>>>> 35821a5794e434117b14797f010602c8e412b36c
```
Remove this.

Salompas replied:
Done.

TomAugspurger commented on pandas/_libs/parsers.pyx:
```diff
@@ -644,6 +646,7 @@ cdef class TextReader:
                     raise ValueError('Multiple files found in compressed '
                                      'zip file %s', str(zip_names))
             elif self.compression == 'xz':
+                if lzma is None: raise RuntimeError("lzma module not available.")
```
Formatting: move the `raise` to its own line.

TomAugspurger replied:
Can you also note in the error message that the user may need to rebuild or reinstall Python with the proper dependencies to fix this.

Salompas replied:
Do you want this message on the warning and on the RuntimeError?

TomAugspurger commented on pandas/util/testing.py:
```diff
@@ -264,9 +268,11 @@ def write_to_compressed(compression, path, data, dest="test"):
-        import lzma
-
-        compress_method = lzma.LZMAFile
+        lzma = import_lzma()
```
This is already done at the top of the file. Can be removed I think.

Salompas commented:
> @Salompas it's an isort issue: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=15881&view=logs&jobId=435ca956-9126-505f-566f-fff31072e2ba&taskId=e3717827-2a2f-5148-8ad4-7900d2171f6e&lineStart=37&lineEnd=38&colStart=1&colEnd=1
> 
> https://dev.pandas.io/development/contributing.html#import-formatting

Thanks for the link, I will fix the issue.

guilherme-salome committed:
Fixed isort.

guilherme-salome committed:
Removed remains from a merge.

guilherme-salome committed:
Moved I/O and LZMA bug fix to a separate section.

guilherme-salome committed:
Unecessary import.

guilherme-salome committed:
Updated RuntimeError message to alert user that a re-install might be required.

Salompas commented:
I made the changes you suggested, let me know how it is looking now.

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -160,6 +159,15 @@ Other
+I/O and LZMA
+~~~~~~~~~~~~
+
+- Issue: Some users may unknowingly have an incomplete Python installation, which lacks the `lzma` module from the standard library. In this case, `import pandas` fails due to an `ImportError`; (:issue: `27575`)
```
You don't need the actual "Issue", "Change, ..."  bullets. Just a paragraph describing things.

"Fails" to "failed", since this has changed.

TomAugspurger reviewed (commented):
Looks pretty close. Can you

1. Rename `import_lzma` to `_import_lzma` (`pandas.compat` is supposed to be private, but people still use it)
2. move the `if lzma is None` check to a new method in `pandas/compat/__init__.py` called `_get_lzma_file(lzma)`. If `lzma` is None it raises the RuntimeError. Otherwise, it returns `lzma.LZMAFile`, which we can pass the kwargs to when lzma is not None.
3. Add at least one subprocess-style test that hits the RuntimeError.

guilherme-salome committed:
Moved the check `lzma is None` to a function, which also raises
RuntimeError if the check fails. Updated remainder of the code to use
this function.

guilherme-salome committed:
Removed bulletpoints.

guilherme-salome committed:
Added test that fails at Runtime when lzma module not available. Test
succeeds if lzma module is available.

Salompas commented:
> Looks pretty close. Can you
> 
> 1. Rename `import_lzma` to `_import_lzma` (`pandas.compat` is supposed to be private, but people still use it)
> 2. move the `if lzma is None` check to a new method in `pandas/compat/__init__.py` called `_get_lzma_file(lzma)`. If `lzma` is None it raises the RuntimeError. Otherwise, it returns `lzma.LZMAFile`, which we can pass the kwargs to when lzma is not None.
> 3. Add at least one subprocess-style test that hits the RuntimeError.

Done.
1., 2. - Good idea! Looks way better now.
3. - The test will fail if you do not have `lzma`, but succeeds otherwise.

Salompas commented:
Just for my future reference, on `Emacs` if you have the following in your `init.el`:
```
(progn (add-hook 'before-save-hook 'whitespace-cleanup)
       (add-hook 'before-save-hook (lambda() (delete-trailing-whitespace))))
```
After saving a file the function `whitespace-cleanup` will substitute spaces for tabs whenever possible.
To change the behavior so that tabs are susbtituted for space, add the following to `init.el`:
`(setq-default indent-tabs-mode nil)
`
Now there should not be any issues with `pyx` files.

Salompas commented:
I guess another modification to this code is to show the warning just once (right now it shows up twice), but this might not be worth the time.

auvipy reviewed (approved):
looks great

TomAugspurger commented on pandas/tests/io/test_compression.py:
```diff
@@ -125,3 +127,29 @@ def test_compression_warning(compression_only):
+def test_with_missing_lzma_runtime():
+    """Tests if RuntimeError is hit when calling lzma without
+    having the module available."""
+    code = textwrap.dedent(
```
This test needs to be updated a bit. We need

1. `sys.modules['lzma'] = None` to ensure it's not present.
2. I'd like to test an actual writer, rather than this internal function.
3. I'd like to assert something about the output... Maybe do something like

```python
df = pd.DataFrame()
with pytest.raises(RuntimeError, match="lzma module"):
    df.to_csv('foo.csv', compression="xz")
```

Salompas replied:
I will get this done. Thanks for the suggestion.

TomAugspurger reviewed (commented):
I think a double warning is fine. Don't worry about that.

One request for changes on the second test. Otherwise, this looks great.

jreback commented on pandas/compat/__init__.py:
```diff
@@ -65,3 +66,27 @@ def is_platform_mac():
+        warnings.warn(msg)
+
+
+def _get_lzma_file(lzma):
```
can you add a doc-string to both of these functions.

Salompas replied:
Yes!

jreback commented on pandas/compat/__init__.py:
```diff
@@ -65,3 +66,27 @@ def is_platform_mac():
+
+def _get_lzma_file(lzma):
+    if lzma is None:
+        raise RuntimeError(
```
this seems overly complicated; why don't we just do this with one function? e.g. import in the get_lzma_file and warn there

TomAugspurger replied:
I think we want a warning an import time right? To let people know that their Python install is incomplete?

jreback replied:
I c, I guess this is ok then.

guilherme-salome committed:
Modified runtime test to make direct call to pandas function.

guilherme-salome committed:
Added docstring.

Salompas commented:
@jreback @TomAugspurger 
Thanks for the suggestions. I made the changes to the code.
Could you check and let me know where to improve? Thanks!

TomAugspurger commented:
Thanks @Salompas!

TomAugspurger commented:
Backported in #28012.

rth commented:
I opened https://github.com/pandas-dev/pandas/issues/43461 to re-visit this. I would like to propose to remove the import warning if lzma is missing. Please comment in the linked issue if you disagree.
